### PR TITLE
Performance improvements for FV

### LINKED ISSUE #23162: Do less warehouse queries when performing finite volume calculations
## Reason
Warehouse queries are showing up as significant contributors to runtime as shown in https://github.com/idaholab/moose/pull/22130#issuecomment-1373009387

## Design
One place where I know we could immediately refactor and get better performance is in the chaining of calls of Dirichlet BC related methods in the variable classes. I think we do a lot of extranous warehouse queries there that could be eliminated by adding a pointer to the Dirichlet BC (or nullptr when it doesn't exist) to some return values

## Impact
Faster run-time for FV simulations

### LINKED ISSUE #23220: Investigate possible slow-down of functor system
## Reason

After merging #22130, I have seen valgrind timeouts for the tests disabled in https://github.com/idaholab/moose/pull/23215 and on CIVET next branch testing for a framework functor test (I will put it here if it crops up again).

## Design
Profile the tests timing out in valgrind both before and after the merge of #22130 and see what's different.

## Impact
We want to add features to our code (like automatic porosity jump handling) without sacrificing performance if we can

### PR DESCRIPTION
**Pull request by lindsayad**
After #22130 there were some slow-downs in some functor-based simulations. This brings back all the losses in the flow over sphere example plus a tiny bit more. Changes:

- Cache last warehouse lookup for Dirichlet boundary conditions
- Don't do pre-init evaluations for INSFV because we use on-the-fly evals entirely

### Pre 22130 graph

pre-porous

### Post 22130 graph
post-porous

### With this PR graph
cache-fi-and-dont-preinit

Closes #23220 #23162

### CODE REVIEW

**Comment by lindsayad** [comment]
Lol apparently we still use it for initial conditions

**Comment by lindsayad** [comment]
Lol, eeked one more second with the dof indices "optimization". Nice use of time :-/

Mixing length test now runs about 6% faster than it did pre the porous PR

### pre porous

mixing-pre-porous

### with this PR, no dof indices optimization

mixing-optimized

### with this PR, with dof indices optimization

mixing-optimized-cache-dof-indices

**Comment by GiudGiud** [comment]
nice. Is there a diff option for pprof graphs? Just curious

**Comment by lindsayad** [comment]
> nice. Is there a diff option for pprof graphs? Just curious

There is. I've used it before a while ago

**Comment by lindsayad** [comment]
@grmnptr @GiudGiud this is ready for review

**Comment by grmnptr** [comment]
Got it! Mine is 50% done, will try to finish today!

File: framework/include/variables/MooseVariableBase.h
Comment (grmnptr) [inline framework/include/variables/MooseVariableBase.h:180]: This is cool! Just a quick question: in what cases do we need this for FV at the moment? I was thinking about this lately, if we didn't need thee evaluations we could get rid of most of the `MooseVariableFVData`, right? Or maybe I am wrong here. I suppose when we evaluate with quadrature points we need these.

File: framework/include/variables/MooseVariableDataFV.h
Comment (grmnptr) [inline framework/include/variables/MooseVariableDataFV.h:435]: Can this be inlined? Also, whenever I see friend functions in moose it is really restrictive (like `Controls` and `ControllableParameters` is this going to be similarly restrictive?

File: framework/src/variables/MooseVariableFV.C
Comment (grmnptr) [inline framework/src/variables/MooseVariableFV.C:377]: This is a very nice catch, especially when we have gradient evaluations where we use the value so many times!

File: framework/include/variables/MooseVariableFV.h
Comment (grmnptr) [inline framework/include/variables/MooseVariableFV.h:611]: I think we should go into details here with the reason why we need this.

File: framework/src/variables/MooseVariableFV.C
Comment (grmnptr) [inline framework/src/variables/MooseVariableFV.C:451]: Just a few comments here so people who look at this guy know why we are caching!

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i
Comment (grmnptr) [inline modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i:26]: Is there a way to know if we need this or not? I mean we should provide a guidance for the users so they can decide. Or is this just to make sure the exodiff checks out?

**Review by grmnptr — CHANGES_REQUESTED** [review]
I like it! Just a few questions/comments.

File: framework/src/auxkernels/ParsedAux.C
Comment (GiudGiud) [inline framework/src/auxkernels/ParsedAux.C:62]: what is happening now? are we not pre-computing the values?

File: framework/src/auxkernels/VectorMagnitudeAux.C
Comment (GiudGiud) [inline framework/src/auxkernels/VectorMagnitudeAux.C:38]: any change we can move all this to the base class for auxkernels?
we can check which functors are requested, and request the QpComputations

File: modules/navier_stokes/src/actions/NSFVAction.C
Comment (GiudGiud) [inline modules/navier_stokes/src/actions/NSFVAction.C:276]: can you delete this typedef of CoupledName then

**Review by GiudGiud — COMMENTED** [review]
so if we now init FV data on Qps when requested, can we use FV variables in FE kernels?

File: framework/include/variables/MooseVariableBase.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableBase.h:180]: The fully compressible finite volume kernels are currently using qp indexing

File: framework/include/variables/MooseVariableDataFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableDataFV.h:435]: `initDofIndices` is a template function defined slightly higher up in the diff. Since it is templated it is implicitly marked as `inline`. What do you mean by similarly restrictive? The template function has been declared as a friend of this class so it can access private/protected data members.

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h:612]: I thought about this and went with the current code mostly out of laziness. I will try your suggestion out

File: framework/src/interfaces/Coupleable.C
Comment (lindsayad) [reply framework/src/interfaces/Coupleable.C:252]: I suppose since this is a one-liner I can just write out the code. But if people modify one (const/non-const) they will have to remember to modify the other

File: framework/src/auxkernels/ParsedAux.C
Comment (lindsayad) [reply framework/src/auxkernels/ParsedAux.C:62]: not by default for INSFV/PINSFV/WCNSFV

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i:26]: I think I went through and made sure that (common) qp-consumer objects mark requirements for qp-calculations (see https://github.com/idaholab/moose/pull/23233/files#r1099323717), so I believe these input file modifications can actually be removed. However, if someone has forgotten to mark a consumer as needing qp calculations, then this input file option could be used to accomplish the same purpose.

File: framework/include/interfaces/Coupleable.h
Comment (lindsayad) [reply framework/include/interfaces/Coupleable.h:1422]: This is not necessarily an array variable. People can pass a multiple standard field variables

File: framework/src/auxkernels/VectorMagnitudeAux.C
Comment (lindsayad) [reply framework/src/auxkernels/VectorMagnitudeAux.C:38]: This isn't a functor though in this derived class. And many aux kernels now have functor versions which do not need qp computations. You could be registering false positives for needing qp computations and incurring computational cost that is not needed if you just made all coupled variables in every aux kernel `requireQpComputations`

Comment (GiudGiud) [reply framework/src/auxkernels/VectorMagnitudeAux.C:38]: yeah I suppose I worded this wrong. 

We can check which finite volume variables / other objects needing explicit Qp computations are being queried through a non-functor interface. Then we can requireQpComputations in the base class ?

File: framework/include/variables/MooseVariableFV.h
Comment (GiudGiud) [reply framework/include/variables/MooseVariableFV.h:612]: instead rather than as well?

**Comment by lindsayad** [comment]
Oh wow @grmnptr such beautiful code suggestion. Converting to a map constructed during `initialSetup` and avoiding doing any "run-time" queries leads to lopping off another 7 seconds of simulation time from 93 seconds for the previous best of the twice refined mixing length channel test to a new best of 86 seconds!

use-map

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h]: Done. Excellent suggestion!

File: framework/include/variables/MooseVariableFV.h
Comment (lindsayad) [reply framework/include/variables/MooseVariableFV.h]: Removed this cache variable in favor of non-mutable map constructed during `initialSetup`

File: framework/src/variables/MooseVariableFV.C
Comment (lindsayad) [reply framework/src/variables/MooseVariableFV.C]: Removed this cache variable in favor of non-mutable map constructed during `initialSetup`

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i:26]: I will think more about whether there is a fool-proof way to detect whether qp-computations are required in the library per @GiudGiud's suggestion (https://github.com/idaholab/moose/pull/23233#discussion_r1100863857), and if so, then I can remove the input file option totally

File: modules/navier_stokes/src/actions/NSFVAction.C
Comment (lindsayad) [reply modules/navier_stokes/src/actions/NSFVAction.C:276]: Hmm there are still a decent number of places in (P)INSFV and consequently in the action where we use coupled variables

Comment (lindsayad) [reply modules/navier_stokes/src/actions/NSFVAction.C:276]: Changing that could be more follow on work

**Comment by lindsayad** [comment]
After refactoring all of (P)INSFV to not use `addCoupledVar` but instead use `addParam<MooseFunctorName>` we are even faster (although this could just be statistical noise)

post-review

File: framework/src/auxkernels/VectorMagnitudeAux.C
Comment (lindsayad) [reply framework/src/auxkernels/VectorMagnitudeAux.C]: Alright I've updated `Coupleable` such that if a finite volume variable is requested through `add.*CoupledVar` then we `requireQpComputations`. So someone wishing to not trigger pre-init'd qp calculations should use `addParam<MooseFunctorName>`

File: modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i
Comment (lindsayad) [reply modules/navier_stokes/test/tests/finite_volume/ins/lid-driven/lid-driven.i]: Ok I was able to delete the `qp_computations` parameter

File: modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C
Comment (GiudGiud) [inline modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:47]: you should retrieve the skewness argument from the variable?

File: modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C
Comment (GiudGiud) [inline modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: same?

File: modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C
Comment (GiudGiud) [reply modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:47]: I guess we dont have the variables anymore. You d have to query the variable from intialSetup and keep the skewness argument in a class member

File: modules/navier_stokes/src/fvkernels/INSFVMixingLengthScalarDiffusion.C
Comment (GiudGiud) [inline modules/navier_stokes/src/fvkernels/INSFVMixingLengthScalarDiffusion.C:61]: so when we make an argument like this, it defaults to no skewness correction right?
but what if the variable has skewness-correction on, does that bring it back?

File: modules/navier_stokes/src/fvkernels/INSFVMomentumPressure.C
Comment (GiudGiud) [inline modules/navier_stokes/src/fvkernels/INSFVMomentumPressure.C:29]: that works but do we want this on every kernel and on every variable?

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C
Comment (lindsayad) [reply modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:47]: I think @grmnptr expressed interest in having this be controlled more on an operator level as opposed to a global level, but I could be mis-remembering

File: modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C
Comment (lindsayad) [reply modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: we didn't do this before. Now you want to add it now that I've removed the variables?

File: modules/navier_stokes/src/fvkernels/INSFVMixingLengthScalarDiffusion.C
Comment (lindsayad) [reply modules/navier_stokes/src/fvkernels/INSFVMixingLengthScalarDiffusion.C:61]: No it doesn't. `adGradSln` takes a `correct_skewness` argument which defaults to `false`, so the change here preserves previous behavior

File: modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C
Comment (GiudGiud) [reply modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: seems like we should have the option to be consistently using skewness-correction everywhere

Comment (lindsayad) [reply modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: you can open an issue for it. Feature addition/control is not the aim of this PR

Comment (GiudGiud) [reply modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: For sure, but we need to agree that we ll do skewness correction at the kernel level if we re removing access to the variable.

It makes sense to realize now that we re not doing skewness-correction in all the operators as expected

Comment (GiudGiud) [reply modules/navier_stokes/src/fvbcs/INSFVMassAdvectionOutflowBC.C:57]: #23438

File: modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C
Comment (GiudGiud) [reply modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:47]: I ll wait on Peter but essentially I m at an approval point here

Comment (grmnptr) [reply modules/navier_stokes/src/fvbcs/INSFVWallFunctionBC.C:47]: Yes I'll do these once I get to the point where unify limiting/interpolation. I think it would be best to do it on a per-kernel basis.

**Review by GiudGiud — APPROVED** [review]

**Review by grmnptr — APPROVED** [review]

**Pull request merged**
